include/swift/Reflection/ReflectionContext.h

```diff
@@ -644,6 +645,46 @@ class ReflectionContext
     return false;
   }
 
+  /// Adds an image using the FindSection closure to find the swift metadata
+  /// sections. \param[in] FindSection
+  ///     Closure that finds sections by name. ReflectionContext is in charge
+  ///     of freeing the memory buffer in the RemoteRef return value.
+  ///     process.
+  /// \return
+  ///     /b True if any of the reflection sections were registered,
+  ///     /b false otherwise.
+  bool
+  addImage(std::function<std::pair<RemoteRef<void>, uint64_t>(llvm::StringRef)>
+               FindSection) {
+    auto &ObjectFileFormat = SwiftObjectFileFormat::getSwiftObjectFileFormat();
+    auto Sections = {
+        ReflectionSectionKind::fieldmd, ReflectionSectionKind::assocty,
+        ReflectionSectionKind::builtin, ReflectionSectionKind::capture,
+        ReflectionSectionKind::typeref, ReflectionSectionKind::reflstr};
+
+    std::vector<std::pair<RemoteRef<void>, uint64_t>> Pairs;
+    for (auto Section : Sections) {
+      Pairs.push_back(FindSection(ObjectFileFormat.getSectionName(Section)));
+      if (Pairs.back().first) {
+        MemoryReader::ReadBytesResult Buffer(
+            Pairs.back().first.getLocalBuffer(),
+            [](const void *Ptr) { free(const_cast<void *>(Ptr)); });
+
+        savedBuffers.push_back(std::move(Buffer));
+      }
+    }
+
+    if (llvm::all_of(Pairs, [](const auto &Pair) { return !Pair.first; }))
```

Comment:
I think this could benefit from a comment: Are we requiring that all sections are present here?

Reply:
Also, perhaps this could just be an early-exit in the loop above?

Reply:
It requires that at least one section be present, so it can't be an early exit

Reply:
Makes sense now. I stopped reading after `all_of`.